x/ibchooks/marker_hooks.go

```diff
@@ -0,0 +1,235 @@
+package ibchooks
+
+import (
+	"encoding/json"
+	"strconv"
+	"strings"
+
+	sdkerrors "cosmossdk.io/errors"
+
+	sdktypes "github.com/cosmos/cosmos-sdk/types"
+	authtypes "github.com/cosmos/cosmos-sdk/x/auth/types"
+	banktypes "github.com/cosmos/cosmos-sdk/x/bank/types"
+	transfertypes "github.com/cosmos/ibc-go/v6/modules/apps/transfer/types"
+	"github.com/cosmos/ibc-go/v6/modules/core/exported"
+	ibckeeper "github.com/cosmos/ibc-go/v6/modules/core/keeper"
+	tendermintclient "github.com/cosmos/ibc-go/v6/modules/light-clients/07-tendermint/types"
+
+	"github.com/provenance-io/provenance/x/ibchooks/types"
+	markerkeeper "github.com/provenance-io/provenance/x/marker/keeper"
+	markertypes "github.com/provenance-io/provenance/x/marker/types"
+)
+
+type MarkerHooks struct {
+	MarkerKeeper *markerkeeper.Keeper
+}
+
+func NewMarkerHooks(markerkeeper *markerkeeper.Keeper) MarkerHooks {
+	return MarkerHooks{
+		MarkerKeeper: markerkeeper,
+	}
+}
+
+// ProperlyConfigured returns false when marker hooks are configured incorrectly
+func (h MarkerHooks) ProperlyConfigured() bool {
+	return h.MarkerKeeper != nil
+}
+
+// AddUpdateMarker will add or update ibc Marker with transfer authorities
+func (h MarkerHooks) AddUpdateMarker(ctx sdktypes.Context, packet exported.PacketI, ibcKeeper *ibckeeper.Keeper) error {
+	var data transfertypes.FungibleTokenPacketData
+	if err := json.Unmarshal(packet.GetData(), &data); err != nil {
+		return err
+	}
+	ibcDenom := MustExtractDenomFromPacketOnRecv(packet)
+	if !strings.HasPrefix(ibcDenom, "ibc/") {
+		return nil
+	}
+
+	markerAddress, err := markertypes.MarkerAddress(ibcDenom)
+	if err != nil {
+		return err
+	}
+	marker, err := h.MarkerKeeper.GetMarker(ctx, markerAddress)
+	if err != nil {
+		return err
+	}
+	var transferAuthAddrs []sdktypes.AccAddress
+	transferAuthAddrs, coinType, err := ProcessMarkerMemo(data.GetMemo())
+	if err != nil {
+		return err
+	}
+
+	if marker != nil {
+		if err = ResetMarkerAccessGrants(transferAuthAddrs, marker); err != nil {
+			return err
+		}
+		h.MarkerKeeper.SetMarker(ctx, marker)
+		return nil
+	}
+	return h.createNewIbcMarker(data, ibcDenom, coinType, transferAuthAddrs, ctx, packet, ibcKeeper)
+}
+
+// createNewIbcMarker creates a new marker account for ibc token
+func (h MarkerHooks) createNewIbcMarker(data transfertypes.FungibleTokenPacketData, ibcDenom string, coinType markertypes.MarkerType, transferAuthAddrs []sdktypes.AccAddress, ctx sdktypes.Context, packet exported.PacketI, ibcKeeper *ibckeeper.Keeper) error {
+	amount, err := strconv.ParseInt(data.Amount, 10, 64)
+	if err != nil {
+		return err
+	}
+	marker := markertypes.NewMarkerAccount(
+		authtypes.NewBaseAccountWithAddress(markertypes.MustGetMarkerAddress(ibcDenom)),
+		sdktypes.NewInt64Coin(ibcDenom, amount),
+		nil,
+		nil,
+		markertypes.StatusActive,
+		coinType,
+		false, // supply fixed
+		false, // allow gov
+		false, // allow force transfer
```

Comment:
Bringing this out of a slack conversation. Would it be possible to reflect the `allow_forced_transfer` from the source chain to the destination? Figure/FES has some use cases around that.

Reply:
Yes, I can have the source chain add the force transfer flag along with the transfer auths.

Reply:
Done and ready to go.